Fix "Implement Members" action: prefer PSI to `Ty`s when generating PSI

vlad20012 opened the pull request:
Fixes #4967, fixes #5044, fixes #4979, fixes #6340, fixes #2982, fixes #5592, fixes #6085

1. Never expand type aliases (#4967, #5044)
2. Don't expand macro types
3. Auto-import trait names from type bounds (#2982)
4. Resolve name conflicts during auto-import (#4979, #6340):
   1. If the target item is in the scope with a different name (`use A as B`), use that name (#5592, #6085)
   2. If the name exists in the scope, but refers to another item, use a qualified path to the target item
   3. If the target can't be imported, use crate-qualified path to the item (produces incorrect code, but it should be simpler to fix)

changelog:

Kobzol commented:
This is awesome! :rocket: I know that this started with type aliases, but point 4) is actually maybe even better :laughing:

This also fixes https://github.com/intellij-rust/intellij-rust/issues/6340, which is basically a duplicate of https://github.com/intellij-rust/intellij-rust/issues/4979.

I found a few more cases where the type was rendered instead of the type reference being used. I changed them [here](https://github.com/Kobzol/intellij-rust/commit/a214d3d8dc8535901212afdf9721ca7c3f929443). Basically uses of `renderInsertionSafe` are now potentially suspicious :)

BTW, when I run CLion with this PR, I always get `java.lang.RuntimeException: Broken stub format, most likely version of FILE (Rust.file) was not updated after serialization changes` on startup.

Kobzol commented:
Btw, would it still make sense after this PR to move `aliasedBy` from `TyAdt` to `Ty` to better support scenarios where we only have a type from type inference?

vlad20012 commented:
> BTW, when I run CLion with this PR, I always get `java.lang.RuntimeException: Broken stub format, most likely version of FILE (Rust.file) was not updated after serialization changes` on startup.

Hmm, I can't reproduce. Maybe previously you were on some branch with `STUB_VERSION = 215`? 

> I found a few more cases where the type was rendered instead of the type reference being used. I changed them [here](https://github.com/Kobzol/intellij-rust/commit/a214d3d8dc8535901212afdf9721ca7c3f929443). Basically uses of `renderInsertionSafe` are now potentially suspicious :)

:+1: but usage of `substAndGetText` with `Substitution` is also suspicious. Aliases from a `Substitution` are still expanded, for example in
```rust
type Alias = i32;
struct Bar<T> { f: T }
impl Bar<Alias> {}
```

I've introduced `PsiSubstitution` to solve this. You can obtain it from a path with `pathPsiSubst()` function (in the case of impl for struct, the path is `impl.typeReference?.skipParens() as? RsBaseType)?.path`)

> Btw, would it still make sense after this PR to move `aliasedBy` from `TyAdt` to `Ty` to better support scenarios where we only have a type from type inference?

Yes!

Kobzol commented:
>Hmm, I can't reproduce. Maybe previously you were on some branch with STUB_VERSION = 215?

Indeed I was. Probably a false alarm.

> but usage of substAndGetText with Substitution is also suspicious. Aliases from a Substitution are still expanded, for example in

I see. Will you include the changes in getter/setter in this PR or do you want me to open another PR on top of this one where getter/setter generation is fixed?

mchernyavsky commented on src/main/kotlin/org/rust/ide/refactoring/implementMembers/impl.kt:
```diff
@@ -154,82 +180,39 @@ private fun createExtraWhitespacesAroundFunction(left: PsiElement, right: PsiEle
+                is RsTypeAlias ->
+                    "    type ${it.escapedName} = ();"
+                is RsFunction ->
+                    "    ${renderFunctionSignature(it) ?: ""}{\n        todo!()\n    }"
```
`?: ""` is not needed.

mchernyavsky commented on src/main/kotlin/org/rust/ide/refactoring/implementMembers/impl.kt:
```diff
@@ -154,82 +180,39 @@ private fun createExtraWhitespacesAroundFunction(left: PsiElement, right: PsiEle
-                "    const ${it.nameLikeElement.text}: ${it.typeReference?.substAndGetText(subst) ?: "_"} = ${initialValue.text};"
+private class MembersGenerator(
+    private val factory: RsPsiFactory,
+    private val impl: RsImplItem,
```
`val` is not needed.

mchernyavsky commented on src/main/kotlin/org/rust/lang/core/psi/ext/RsFnPointerType.kt:
```diff
@@ -6,7 +6,27 @@
 package org.rust.lang.core.psi.ext
 
 import org.rust.lang.core.psi.RsFnPointerType
+import org.rust.lang.core.psi.RsFunction
```
Unused

mchernyavsky commented on src/main/kotlin/org/rust/ide/presentation/RsPsiRenderer.kt:
```diff
@@ -0,0 +1,715 @@
+import org.rust.ide.utils.import.ImportCandidate
+import org.rust.ide.utils.import.ImportCandidatesCollector
+import org.rust.ide.utils.import.ImportContext
+import org.rust.ide.utils.import.RsImportHelper
```
Unused

mchernyavsky commented on src/main/kotlin/org/rust/ide/presentation/RsPsiRenderer.kt:
```diff
@@ -0,0 +1,715 @@
+    renderLifetimes: Boolean
+): String = buildString { appendFunctionSignature(this, fn, renderLifetimes) }
+
+data class PsiRenderingOptions(
```
Unused

mchernyavsky reviewed (approved):
LGTM except for minor comments

vlad20012 commented:
> Will you include the changes in getter/setter in this PR or do you want me to open another PR on top of this one where getter/setter generation is fixed?

I'd prefer you. If it doesn't bother you...

vlad20012 commented:
:rocket: :rocket: :rocket: 
bors r+

Kobzol commented:
> I'd prefer you. If it doesn't bother you...

No problem, I'll send it right after this PR gets merged :)
